### ISSUE TITLE
safeguards have to refined for getting sources when running in JDT

### ISSUE THREAD

**Issue by elucash** [immutables/immutables#259]
```
ENTRY org.eclipse.jdt.apt.pluggable.core 1 1 2016-02-17 17:09:19.723
!MESSAGE org.immutables.value.internal.$processor$.$Processor threw java.lang.IllegalArgumentException: Unsupported location: SOURCE_PATH
        at org.eclipse.jdt.internal.apt.pluggable.core.filer.IdeFilerImpl.getFileFromOutputLocation(IdeFilerImpl.java:188)
        at org.eclipse.jdt.internal.apt.pluggable.core.filer.IdeFilerImpl.getResource(IdeFilerImpl.java:154)
        at org.immutables.value.internal.$generator$.$SourceExtraction$1.extract($SourceExtraction.java:141)
        at org.immutables.value.internal.$generator$.$SourceExtraction$CompositeExtractor.extract($SourceExtraction.java:255)
        at org.immutables.value.internal.$generator$.$SourceExtraction.extract($SourceExtraction.java:127)
        at org.immutables.value.internal.$generator$.$SourceExtraction.readImports($SourceExtraction.java:111)
        at org.immutables.value.internal.$processor$.meta.$Proto$DeclaringType.sourceImports($Proto.java:657)
        at org.immutables.value.internal.$processor$.meta.$ImmutableProto$DeclaringType.sourceImports($ImmutableProto.java:475)
        at org.immutables.value.internal.$processor$.meta.$TypeStringProvider.getFromSourceImports($TypeStringProvider.java:160)
        at org.immutables.value.internal.$processor$.meta.$TypeStringProvider.resolveIfPossible($TypeStringProvider.java:142)
        at org.immutables.value.internal.$processor$.meta.$TypeStringProvider.appendResolved($TypeStringProvider.java:127)
        at org.immutables.value.internal.$processor$.meta.$TypeStringProvider.caseType($TypeStringProvider.java:236)
        at org.immutables.value.internal.$processor$.meta.$TypeStringProvider.appendTypeArguments($TypeStringProvider.java:293)
        at org.immutables.value.internal.$processor$.meta.$TypeStringProvider.caseType($TypeStringProvider.java:237)
        at org.immutables.value.internal.$processor$.meta.$TypeStringProvider.process($TypeStringProvider.java:101)
        at org.immutables.value.internal.$processor$.meta.$ValueAttribute.initTypeName($ValueAttribute.java:852)
        at org.immutables.value.internal.$processor$.meta.$ValueAttribute.initAndValidate($ValueAttribute.java:814)
        at org.immutables.value.internal.$processor$.meta.$AccessorAttributesCollector.collect($AccessorAttributesCollector.java:94)
        at org.immutables.value.internal.$processor$.meta.$ValueTypeComposer.compose($ValueTypeComposer.java:108)
        at org.immutables.value.internal.$processor$.meta.$Round.collectValues($Round.java:112)
        at org.immutables.value.internal.$processor$.$Processor.process($Processor.java:48)
        at org.immutables.value.internal.$generator$.$AbstractGenerator.process($AbstractGenerator.java:89)
        at org.eclipse.jdt.internal.compiler.apt.dispatch.RoundDispatcher.handleProcessor(RoundDispatcher.java:139)
        at org.eclipse.jdt.internal.compiler.apt.dispatch.RoundDispatcher.round(RoundDispatcher.java:121)
        at org.eclipse.jdt.internal.compiler.apt.dispatch.BaseAnnotationProcessorManager.processAnnotations(BaseAnnotationProcessorManager.java:159)
        at org.eclipse.jdt.internal.apt.pluggable.core.dispatch.IdeAnnotationProcessorManager.processAnnotations(IdeAnnotationProcessorManager.java:134)
        at org.eclipse.jdt.internal.compiler.Compiler.processAnnotations(Compiler.java:967)
        at org.eclipse.jdt.internal.compiler.Compiler.compile(Compiler.java:489)
        at org.eclipse.jdt.internal.core.builder.AbstractImageBuilder.compile(AbstractImageBuilder.java:395)
        at org.eclipse.jdt.internal.core.builder.BatchImageBuilder.compile(BatchImageBuilder.java:183)
        at org.eclipse.jdt.internal.core.builder.AbstractImageBuilder.compile(AbstractImageBuilder.java:332)
        at org.eclipse.jdt.internal.core.builder.BatchImageBuilder.build(BatchImageBuilder.java:62)
        at org.eclipse.jdt.internal.core.builder.JavaBuilder.buildAll(JavaBuilder.java:256)
        at org.eclipse.jdt.internal.core.builder.JavaBuilder.build(JavaBuilder.java:175)
        at org.eclipse.core.internal.events.BuildManager$2.run(BuildManager.java:734)
        at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
        at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:205)
        at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:245)
        at org.eclipse.core.internal.events.BuildManager$1.run(BuildManager.java:300)
        at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
        at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:303)
        at org.eclipse.core.internal.events.BuildManager.basicBuildLoop(BuildManager.java:359)
        at org.eclipse.core.internal.events.BuildManager.build(BuildManager.java:382)
        at org.eclipse.core.internal.events.AutoBuildJob.doBuild(AutoBuildJob.java:144)
        at org.eclipse.core.internal.events.AutoBuildJob.run(AutoBuildJob.java:235)
        at org.eclipse.core.internal.jobs.Worker.run(Worker.java:55)
```

**Comment by elucash** [comment]
Released in 2.1.11

**Comment by gilbode** [comment]
With 2.1.11 - the issue does seem resolved for interfaces.  However, changing the hierarchy to abstract class (which was my original working workaround) no longer works and generates this stack trace:

!MESSAGE org.immutables.value.internal.$processor$.$Processor threw java.lang.IllegalArgumentException
	at org.immutables.value.internal.$guava$.base.$Preconditions.checkArgument($Preconditions.java:108)
	at org.immutables.value.internal.$generator$.$TypeHierarchyCollector.toDeclaredType($TypeHierarchyCollector.java:45)
	at org.immutables.value.internal.$generator$.$TypeHierarchyCollector.collectHierarchyMirrors($TypeHierarchyCollector.java:82)
	at org.immutables.value.internal.$generator$.$TypeHierarchyCollector.collectFrom($TypeHierarchyCollector.java:40)
	at org.immutables.value.internal.$processor$.meta.$TypeIntrospectionBase.collectTypeHierarchy($TypeIntrospectionBase.java:127)
	at org.immutables.value.internal.$processor$.meta.$ValueType.collectTypeHierarchy($ValueType.java:949)
	at org.immutables.value.internal.$processor$.meta.$TypeIntrospectionBase.intospectTypeMirror($TypeIntrospectionBase.java:120)
	at org.immutables.value.internal.$processor$.meta.$TypeIntrospectionBase.introspectType($TypeIntrospectionBase.java:116)
	at org.immutables.value.internal.$processor$.meta.$TypeIntrospectionBase.ensureTypeIntrospected($TypeIntrospectionBase.java:70)
	at org.immutables.value.internal.$processor$.meta.$ValueType.extendedClasses($ValueType.java:954)
	at org.immutables.value.internal.$processor$.meta.$ValueTypeComposer.checkForTypeHierarchy($ValueTypeComposer.java:194)
	at org.immutables.value.internal.$processor$.meta.$ValueTypeComposer.compose($ValueTypeComposer.java:105)
	at org.immutables.value.internal.$processor$.meta.$Round.collectValues($Round.java:112)
	at org.immutables.value.internal.$processor$.$Processor.process($Processor.java:48)
	at org.immutables.value.internal.$generator$.$AbstractGenerator.process($AbstractGenerator.java:89)
	at org.eclipse.jdt.internal.compiler.apt.dispatch.RoundDispatcher.handleProcessor(RoundDispatcher.java:139)
	at org.eclipse.jdt.internal.compiler.apt.dispatch.RoundDispatcher.round(RoundDispatcher.java:121)
	at org.eclipse.jdt.internal.compiler.apt.dispatch.BaseAnnotationProcessorManager.processAnnotations(BaseAnnotationProcessorManager.java:159)
	at org.eclipse.jdt.internal.apt.pluggable.core.dispatch.IdeAnnotationProcessorManager.processAnnotations(IdeAnnotationProcessorManager.java:134)
	at org.eclipse.jdt.internal.compiler.Compiler.processAnnotations(Compiler.java:967)
	at org.eclipse.jdt.internal.compiler.Compiler.compile(Compiler.java:489)
	at org.eclipse.jdt.internal.core.builder.AbstractImageBuilder.compile(AbstractImageBuilder.java:395)
	at org.eclipse.jdt.internal.core.builder.IncrementalImageBuilder.compile(IncrementalImageBuilder.java:331)
	at org.eclipse.jdt.internal.core.builder.AbstractImageBuilder.compile(AbstractImageBuilder.java:332)
	at org.eclipse.jdt.internal.core.builder.IncrementalImageBuilder.build(IncrementalImageBuilder.java:136)
	at org.eclipse.jdt.internal.core.builder.JavaBuilder.buildDeltas(JavaBuilder.java:267)
	at org.eclipse.jdt.internal.core.builder.JavaBuilder.build(JavaBuilder.java:195)
	at org.eclipse.core.internal.events.BuildManager$2.run(BuildManager.java:734)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:205)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:245)
	at org.eclipse.core.internal.events.BuildManager$1.run(BuildManager.java:300)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:303)
	at org.eclipse.core.internal.events.BuildManager.basicBuildLoop(BuildManager.java:359)
	at org.eclipse.core.internal.events.BuildManager.build(BuildManager.java:382)
	at org.eclipse.core.internal.events.AutoBuildJob.doBuild(AutoBuildJob.java:144)
	at org.eclipse.core.internal.events.AutoBuildJob.run(AutoBuildJob.java:235)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:55)

**Comment by elucash** [comment]
I have some ideas, but working mostly blindly here )  Can you, please, either create minimal repro or describe structure of classes/interfaces in detail and how they distributed among modules. And how eclipse eventually reference modules (if we go past by some gradle/or maven integration layer we will still end up with library added as as source project or some jar). Also interesting what exact workaround worked before it was "fixed" last time?

**Comment by gilbode** [comment]
OK, I'm able to create a reproducer, only requires a single project.  This project has nothing but immutables on the classpath / factorypath.  This compiles with 2.1.10 and fails with 2.1.11:

```
public abstract class GenericAbstractClassBaseType<T> {
  public static abstract class Builder<V extends GenericAbstractClassBaseType<?>> {
    public abstract Builder<V> from(V other);
    public abstract GenericAbstractClassBaseType<?> build();
  }
}
```

```
import org.immutables.value.Value;
@Value.Immutable
public abstract class AbstractMyType implements SomeInterface {
}
```

```
import org.immutables.value.Value;
@Value.Immutable
public abstract class AbstractMyTypePagedResult extends GenericAbstractClassBaseType<ImmutableMyType> {
  public abstract static class Builder
      extends GenericAbstractClassBaseType.Builder<AbstractMyTypePagedResult> {
  }
}
```

```
public interface SomeInterface {
}
```

**Comment by gilbode** [comment]
OK I'm able to reproduce with a reproducer - reproducible in a single
eclipse module.  I've updated the github issue w/ the examples.  The
"workaround" was to change the generic base type from interface to abstract
class.

**Comment by elucash** [comment]
Wow, this is hardcore usage! It confirms my suspicions about `extends GenericAbstractClassBaseType<ImmutableMyType>` here we have not a `DECLARED_TYPE` as a supertype, but some sort of `ERROR_TYPE` as `ImmutableMyType` type argument is unresolved because might not be generated yet. I'll play a bit, but sure it can be fixed. Thank you for narrowing it down!

**Comment by elucash** [comment]
Fixed in 384b57109ed8cdd472159b19f26989f16ae6e620  Will be released shortly. Thanks!

**Comment by elucash** [comment]
This should largely work now with both abstract classes and interfaces in 2.1.12.
Let us know if it doesn't!
